reaper.go

```diff
@@ -0,0 +1,184 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+
+	"github.com/opencontainers/runc/libcontainer"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type reaper struct {
+	sync.RWMutex
+	sync.WaitGroup
+
+	exitCodeChans map[int]chan int
+}
+
+func exitStatus(status unix.WaitStatus) int {
+	if status.Signaled() {
+		return exitSignalOffset + int(status.Signal())
+	}
+
+	return status.ExitStatus()
+}
+
+func (r *reaper) getExitCodeCh(pid int) (chan int, error) {
+	r.RLock()
+	defer r.RUnlock()
+
+	exitCodeCh, exist := r.exitCodeChans[pid]
+	if !exist {
+		return nil, fmt.Errorf("Process %d not found", pid)
+	}
+
+	return exitCodeCh, nil
+}
+
+func (r *reaper) setExitCodeCh(pid int, exitCodeCh chan int) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.exitCodeChans[pid] = exitCodeCh
+}
+
+func (r *reaper) deleteExitCodeCh(pid int) {
+	r.Lock()
+	defer r.Unlock()
+
+	exitCodeCh, exist := r.exitCodeChans[pid]
+	if !exist {
+		return
+	}
+
+	close(exitCodeCh)
+
+	delete(r.exitCodeChans, pid)
+}
+
+func (r *reaper) reap() error {
+	var (
+		ws  unix.WaitStatus
+		rus unix.Rusage
+	)
+
+	// When running new processes, libcontainer expects to wait
+	// for the first process actually spawning the container.
+	// This lock allows any code starting a new process to take
+	// the lock prior to the start of this new process, preventing
+	// the subreaper from reaping unwanted processes.
+	r.Wait()
```

Comment:
A new request might just arrive after r.Wait() returns.
In this case the new call into the libcontainer and the next unix.Wait4() happen parallel.
One of the methods to solve this problem is changing the sync.WaitGroup to sync.RWMutex too.

Reply:
Yes I agree. But to make this safe, I will have to lock the entire `for` loop, right ? This is the only way to make sure the new request will wait until the reaper exits the loop.

Reply:
PR updated!